Issue 27019/doc (#29434)
airbytehq/airbyte: airbyte-cdk/python/airbyte_cdk/sources/file_based/README.md

### Parquet
Parquet is a file format defined by [Apache](https://parquet.apache.org/). Configuration options are:
* `decimal_as_float`: Whether to convert decimal fields to floats. There is a loss of precision when converting decimals to floats, so this is not recommended.

## Schema

Having a schema allows for the file-based CDK to take action when there is a discrepancy between a record and what are the expected types of the record fields.

Schema can be either inferred or user provided.
* If the user defines it a format using JSON types, inference will not apply. Input schemas are a key/value pair of strings describing column name and data type. Supported types are `["string", "number", "integer", "object", "array", "boolean", "null"]`. For example, `{"col1": "string", "col2": "boolean"}`.

> * If the user enables schemaless sync, schema will `{"data": "object"}` and therefore emitted records will look like `{"data": {"col1": val1, …}}`.

Reviewer: ```suggestion
* If the user enables schemaless sync, schema will `{"data": "object"}` and therefore emitted records will look like `{"data": {"col1": val1, …}}`. This is recommended if the contents between files in the stream vary significantly, and/or if data is very nested.
```

> * Else, the file-based CDK will infer the schema depending on the file type. Some file formats defined the schema as part of their metadata (like Parquet), some do on the record-level (like Avro) and some don't have any explicit typing (like JSON or CSV). Any file format that does not define their schema on a metadata level will require the file-based CDK to iterate to a number of records. There is a limit of bytes that will be consumed in order to infer the schema.

Reviewer: ```suggestion
* Else, the file-based CDK will infer the schema depending on the file type. Some file formats defined the schema as part of their metadata (like Parquet), some do on the record-level (like Avro) and some don't have any explicit typing (like JSON or CSV). Note that all CSV values are inferred as strings except where we are supporting legacy configurations. Any file format that does not define their schema on a metadata level will require the file-based CDK to iterate to a number of records. There is a limit of bytes that will be consumed in order to infer the schema.
```

### Validation Policies
Users will be required to select one of 3 different options, in the event that records are encountered that don’t conform to the schema.

> * Skip nonconforming records: check each record to see if it conforms to the schema; keep a count of the number of records in each file that do and do not conform and emit a log message with these counts once we’re done reading the file. Do not emit records that don’t conform.

Reviewer: ```suggestion
* Skip nonconforming records: check each record to see if it conforms to the user-input or inferred schema; skip the record if it doesn't conform. We keep a count of the number of records in each file that do and do not conform and emit a log message with these counts once we’re done reading the file.
```

> * Emit all records: columns that don’t exist in the configured catalog may still be available in the destination’s raw tables, but this is undefined behavior.

Reviewer: ```suggestion
* Emit all records: emit all records, even if they do not conform to the user-provided or inferred schema. Columns that don’t exist in the configured catalog may still be available in the destination’s raw tables, but this is undefined behavior.
```

Reviewer 2: nit: we should say that columns that don't exist in the configured catalog probably won't be available in the destination's table since that's the current behavior. (I agree it should still be considered undefined behavior)

Only error if there are conflicting field types or malformed rows.

> * Stop the sync and wait for schema re-discovery:  if a record is encountered that does not conform to the configured catalog’s schema, we log a message and stop the whole sync.

Reviewer: ```suggestion
* Stop the sync and wait for schema re-discovery:  if a record is encountered that does not conform to the configured catalog’s schema, we log a message and stop the whole sync. Note: this option is not recommended if the files have very different columns or datatypes, because the inferred schema may vary significantly at discover time.
```

When the `schemaless` is enabled, validation will be skipped.

## Breaking Changes (compared to previous S3 implementation)

> TODO * Different behavior if input schema maps to an object or array

Author: Still working on reproducing this one. Also not that I have not documented the breaking changes for `auto_dict_encode` and `timestamp_parsers` as I was not able to reproduce

Reviewer 2: you can use these two commands after pulling the source-s3 secrets:

1. read: `python main.py read --config secrets/legacy_csv_user_schema_cast_complex.json --catalog integration_tests/configured_catalogs/csv.json`
2. `python main.py discover --config secrets/legacy_csv_user_schema_cast_complex.json`

What's interesting is this particular config doesn't work in legacy s3 (unless the destination is able to cast the strings to objects) because the `dict` is discovered as an object, but the it's a string in the record

| Replicate Multiple Streams \(distinct tables\) | Yes        |
| Namespaces                                     | No         |

> We recommend you do not (do not what? the suspense is unbearable!)

Reviewer 2: ```suggestion
We recommend you do not manually modify files that are already synced. The connector has file-level granularity, which means adding or modifying a row in a CSV file will trigger a re-sync of the content of that file.
```

> We recommend you do not (do not what? the suspense is unbearable!)

Reviewer 3: just don't
